### PR TITLE
walletunlocker/lnd: eliminate extra scrypt invocation

### LINKED ISSUE #1223: walletunlocker+lnd: eliminate extra scrypt invocation on wallet unlock
Atm, when we unlock the daemon for the the first time, we incur 3 scrypt invocations: [checking if the wallet password is correct](https://github.com/lightningnetwork/lnd/blob/master/walletunlocker/service.go#L263), [attempting to decrypt the root macaroon store](https://github.com/lightningnetwork/lnd/blob/master/walletunlocker/service.go#L280), and [a third time when we go to create the `LightningWallet` itself](https://github.com/lightningnetwork/lnd/blob/master/chainregistry.go#L514). Only two are absolutely necessary. Getting rid of the extra invocation will speed up the start time, and also lessen the computational/memory load on constrained devices (mobile phones, pi's, etc)

# Steps To Completion 

- [ ] Eliminate the final scrypt invocation. This can be achieved by passing the instantiated `WalletController` implementation directly from the `WalletUnlocker`. This may require a bit of re-working in the way we handle config parsing.

### PR DESCRIPTION
**Pull request by guggero**
This PR fixes #1223:
The number of scrypt invocations is reduced from 3 to 2 by passing the unlocked wallet instance back from the `UnlockerService` when using the `unlock` command.

I wasn't able to see a big decrease in the time needed to unlock the wallet on my fast notebook (around 3.2 seconds, both with and without the improvement).
But as mentioned in the issue, this should be more noticeable on devices with less computing power.

### CODE REVIEW

**Comment by cfromknecht** [comment]
tACK! Tried this locally and time to spin up lnd dropped dramatically. From what I could discern from the logs it looks like it might have taken off about 1.2s :)

Can you split up the commits so there is one per modified file??

**Comment by guggero** [comment]
Cool :sunglasses:
I split the commits as you requested.

File: walletunlocker/service.go [hunk 1]
```diff
@@ -287,6 +287,7 @@ func (u *UnlockerService) UnlockWallet(ctx context.Context,
 	walletUnlockMsg := &WalletUnlockMsg{
 		Passphrase:     password,
 		RecoveryWindow: recoveryWindow,
+		Wallet:         unlockedWallet,
```
Comment (halseth) [inline walletunlocker/service.go]: wallet must be unloaded in case the macaroon check above fails.

File: lnd.go [hunk 2]
```diff
@@ -1011,6 +1020,7 @@ func waitForWalletPassword(
 		walletInitParams := &WalletUnlockParams{
 			Password:       unlockMsg.Passphrase,
 			RecoveryWindow: unlockMsg.RecoveryWindow,
+			Wallet:         unlockMsg.Wallet,
```
Comment (halseth) [inline lnd.go]: can we also avoid unloading the wallet in `case initMsg := <-pwService.InitMsgs:`?

File: walletunlocker/service.go (hunk 1 above)
Comment (guggero) [reply walletunlocker/service.go]: Uh, thanks! I missed this one...

File: lnd.go (hunk 2 above)
Comment (guggero) [reply lnd.go]: Makes a lot of sense! Fixed it.

**Comment by Roasbeef** [comment]
Needs to be rebased after the recent addition to change the encryption passphrase for a user.

**Comment by guggero** [comment]
Rebased :)

**Comment by guggero** [comment]
Sorry, I seem to have missed the comment and long line during the last rebase...
Fixed it now.

File: lnd.go [hunk 3]
```diff
@@ -997,23 +1006,27 @@ func waitForWalletPassword(grpcEndpoints, restEndpoints []string,
...
 		)
 		if err != nil {
-			return nil, err
-		}
+			ltndLog.Errorf("Could not create new wallet: %v", err)
 
-		if err := loader.UnloadWallet(); err != nil {
+			// Don't leave the file open in case the new wallet
+			// could not be created for whatever reason.
+			if err := loader.UnloadWallet(); err != nil {
```
Comment (halseth) [inline lnd.go]: Actually, a better pattern would be 
```go
if err != nil {
    if err := loader.UnloadWallet(); err != nil {
           log(err)
    }
    return nil, err
}
```
as it's the first error that causes us to fail in the first place, and makes sense to return 😛

Comment (guggero) [reply lnd.go]: Ah, that's what you meant... Makes more sense, thanks. I fixed it.

**Comment by Roasbeef** [comment]
Committed as cd8a21c4b829def9bf317a1a9b42f07e36b5891c.

**Pull request closed without merging**